crates/stages/src/stages/merkle.rs

```diff
@@ -156,13 +168,29 @@ impl<DB: Database> Stage<DB> for MerkleStage {
         }
 
         let current_root = tx.get_header(input.stage_progress)?.state_root;
-
         let from_transition = tx.get_block_transition(input.unwind_to)?;
         let to_transition = tx.get_block_transition(input.stage_progress)?;
 
-        let block_root = DBTrieLoader::new(tx.deref_mut())
-            .update_root(current_root, from_transition..to_transition)
-            .map_err(|e| StageError::Fatal(Box::new(e)))?;
+        let mut loader = DBTrieLoader::new(tx.deref_mut());
+        let block_root = loop {
+            match loader
+                .update_root(current_root, from_transition..to_transition)
+                .map_err(|e| StageError::Fatal(Box::new(e)))?
+            {
+                TrieProgress::Complete(root) => break root,
+                TrieProgress::InProgress(_) => {
+                    // Save the loader's progress & drop it
+                    // to allow committing to the database (otherwise we're hitting the borrow
+                    // checker)
+                    let progress = loader.current;
+                    drop(loader);
+                    tx.commit()?;
+                    // Reinstantiate the loader from where it was left off.
+                    loader = DBTrieLoader::new(tx.deref_mut());
+                    loader.current = progress;
+                }
```

Comment:
This is unfortunately necessary, we've already mutably borrowed the transaction to `update_root` and we cannot mutably borrow it again to `commit()`. But calling `loader.tx.commit()` is `mut self` because we use the vanilla DbTxMut method, so it will consume the tx w/o re-opening. 

So instead, we drop the loader, freeing up the `Transaction` object, commit/reopen tx, and then re-instantiate the loader from where it was left off. Should be fine.